internal/framework/fetch/fetch.go

```diff
@@ -0,0 +1,569 @@
+	}
+
+	if result != nil {
+		return result, nil
+	}
+
+	// This case should ideally not be reached, but as a fallback, return the last known error.
+	if lastErr != nil {
+		return nil, lastErr
+	}
+
+	return nil, fmt.Errorf("failed to fetch HTTP file %s: unknown error", targetURL)
+}
+
+// getS3Object fetches an object from S3.
+func (f *DefaultFetcher) getS3Object(
+	ctx context.Context,
+	bucket, key string,
+	timeout time.Duration,
+) ([]byte, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+
+	result, err := f.s3Client.GetObject(ctxWithTimeout, input)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get S3 object: %w", err)
+	}
+	defer result.Body.Close()
+
+	data, err := io.ReadAll(result.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read S3 object body: %w", err)
+	}
+
+	return data, nil
+}
+
+// getHTTPContent fetches content via HTTP(S).
+func (f *DefaultFetcher) getHTTPContent(
+	ctx context.Context,
+	targetURL string,
+	timeout time.Duration,
+) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, targetURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+	req.Header.Set("User-Agent", userAgent)
+
+	resp, err := f.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, &HTTPStatusError{StatusCode: resp.StatusCode}
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read HTTP response body: %w", err)
+	}
+
+	return body, nil
+}
+
+// validateS3FileContent validates the fetched S3 file content using the enabled validation methods.
+func (f *DefaultFetcher) validateS3FileContent(
+	ctx context.Context,
+	data []byte,
+	bucket, key string,
+	opts options,
+) error {
+	if opts.checksumEnabled {
+		if err := f.validateS3Checksum(ctx, data, bucket, key, opts); err != nil {
+			return fmt.Errorf("checksum validation failed: %w", err)
+		}
+	}
+	return nil
+}
+
+// validateS3Checksum validates S3 file content against a SHA256 checksum.
+func (f *DefaultFetcher) validateS3Checksum(
+	ctx context.Context,
+	data []byte,
+	bucket, key string,
+	opts options,
+) error {
+	checksumBucket := bucket
+	checksumKey := key + checksumFileSuffix
+
+	if opts.checksumLocation != "" {
+		if strings.HasPrefix(opts.checksumLocation, "s3://") {
+			// Parse full S3 URL
+			var err error
+			checksumBucket, checksumKey, err = parseS3URL(opts.checksumLocation)
+			if err != nil {
+				return fmt.Errorf("invalid checksum S3 URL: %w", err)
+			}
+		} else {
+			checksumKey = opts.checksumLocation
+		}
+	}
+
+	checksumData, err := f.getS3Object(ctx, checksumBucket, checksumKey, opts.timeout)
```

Comment:
looks like you call `getS3Object` twice. Once at the beginning, then you call validate and call it again here.